sdk/python/kfp/dsl/component_factory.py

```diff
@@ -204,188 +205,187 @@ def _maybe_make_unique(name: str, names: List[str]):
-        elif signature.return_annotation is not None and signature.return_annotation != inspect.Parameter.empty:
-            output_name = _maybe_make_unique(SINGLE_OUTPUT_NAME, output_names)
-            # Fixes exotic, but possible collision:
-            #   `def func(output_path: OutputPath()) -> str: ...`
-            output_names.add(output_name)
-            return_ann = signature.return_annotation
-            if type_annotations.is_list_of_artifacts(return_ann):
-                artifact_cls = return_ann.__args__[0]
-                output_spec = structures.OutputSpec(
-                    type=type_utils.create_bundled_artifact_type(
-                        artifact_cls.schema_title, artifact_cls.schema_version),
-                    is_artifact_list=True)
-            elif type_annotations.is_artifact_class(return_ann):
-                output_spec = structures.OutputSpec(
-                    type=type_utils.create_bundled_artifact_type(
-                        return_ann.schema_title, return_ann.schema_version),
-                    is_artifact_list=False)
-            else:
-                type_struct = type_utils._annotation_to_type_struct(return_ann)
-                output_spec = structures.OutputSpec(type=type_struct)
-
-            outputs[output_name] = output_spec
-    elif return_ann != inspect.Parameter.empty and return_ann != structures.ContainerSpec:
-        raise TypeError(
-            'Return annotation should be either ContainerSpec or omitted for container components.'
+def make_output_spec(annotation: Any) -> structures.OutputSpec:
+    annotation = canonicalize_annotation(annotation)
+    args = make_input_output_spec_args(annotation)
+    return structures.OutputSpec(**args)
+
+
+def make_input_spec(annotation: Any,
+                    inspect_param: inspect.Parameter) -> structures.InputSpec:
+    """Makes an InputSpec from a cleaned output annotation."""
+    annotation = canonicalize_annotation(annotation)
+    input_output_spec_args = make_input_output_spec_args(annotation)
+
+    if (type_annotations.issubclass_of_artifact(annotation) or
+            input_output_spec_args['is_artifact_list']
+       ) and inspect_param.default not in {None, inspect._empty}:
+        raise ValueError(
+            f'Optional Input artifacts may only have default value None. Got: {inspect_param.default}.'
         )
 
-    component_name = name or _python_function_name_to_component_name(
-        func.__name__)
+    default = None if inspect_param.default == inspect.Parameter.empty or type_annotations.issubclass_of_artifact(
+        annotation) else inspect_param.default
+
+    optional = inspect_param.default is not inspect.Parameter.empty or type_utils.is_task_final_status_type(
+        getattr(inspect_param.annotation, '__name__', ''))
+    return structures.InputSpec(
+        **input_output_spec_args,
+        default=default,
+        optional=optional,
+    )
+
+
+def extract_component_interface(
```

Comment:
Note for reviewer: many of the changes in this file refactor the large/complex `extract_component_interface function`, making it much easier to add the logic needed for this PR and maintain this function going forward.